app/tiebout.hoon

```diff
@@ -0,0 +1,253 @@
+  ~&  prize+[wir piz]
+  ?~  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ?+  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ::
+  ::  %our: set config of circle and iterate through messages, sending
+  ::  notifications for all messages where number is higher than our last-read
+  ::
+      {%our @ @}
+    ?>  ?=(%circle -.piz)
+    =/  nom/name:hall  i.t.wir
+    =/  red/@ud  red.loc.cos.piz
+    [~ this(circles.tie.sta (~(put by circles.tie.sta) nom red))]
+  ==
+::
+::  +reap: recieve acknowledgement for peer
+::
+++  reap
+  |=  [wir=wire err=(unit tang)]
+  ^-  (quip move _this)
+  ::~&  reap+[wir =(~ err)]
+  ?~  err
+    ::  XX send message to users inbox
+    [~ this]
+  ?~  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ?+  i.wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+      %our
+    ?<  ?=(~ t.wir)
+    :_  this
+    ~
+  ==
+::
+::  +quit: receive subscription failed, resubscribe
+::
+++  quit
+  |=  wir=wire
+  ^-  (quip move _this)
+  ?~  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ?+  i.wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+      %our
+    ?<  ?=(~ t.wir)
+    :_  this
+    =/  doc/dock  [our.bol %hall]
+    [ost.bol %peer /our/[i.t.wir] doc /circle/[i.t.wir]/config/grams]~
+  ==
+
+::
+::  +diff-hall-rumor: receive message or a read event from a hall circle
+::
+++  diff-hall-rumor
+  |=  [wir=wire rum=rumor:hall]
+  ^-  (quip move _this)
+  ::~&  rumor+[wir rum]
+  ?~  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ?+  wir
+    (mean [leaf+"invalid wire for diff: {(spud wir)}"]~)
+  ::
+  ::  %our
+  ::
+      {%our @ @}
+    ?>  ?=(%circle -.rum)
+    =/  nom/name:hall  i.t.wir
+    ?+  -.rum.rum
+      [~ this]
+    ::
+    ::  %gram: send notification if envelope is lower than read number
+    ::
+      %gram
+    =/  red  (~(get by circles.tie.sta) nom)
+    ?~  red
+      (mean [leaf+"invalid circle for diff: {(spud wir)}"]~)
+    ?:  (gth num.nev.rum.rum u.red)
+      :_  this(circles.tie.sta (~(put by circles.tie.sta) nom u.red))
+      (message-to-notification u.red nev.rum.rum)
+    :_  this
+    (message-to-notification u.red nev.rum.rum)
+    ::
+    ::   %config: set our read number
+    ::
+      %config
+    ?+  -.dif.rum.rum
+      [~ this]
+          %read
+        =/  red/@ud  red.dif.rum.rum
+        [~ this(circles.tie.sta (~(put by circles.tie.sta) nom red))]
+      ==
+    ==
+  ==
+::
+::  generate notification move from hall message
+::
+++  message-to-notification
+  |=  [red=@ud env=envelope:hall]
+  ^-  (list move)
+  ?:  =(aut.gam.env our.bol)
+    ~
+  =/  pay  %-  my  :~
+    alert+s+'New message from {(cite:title aut.gam.env)}'
+  ==
+  =/  not/notification  :+
+    token.tie.sta
+    'com.tlon.urbit-client'
+    pay
+  ?:  (lte num.env red)
+    ~
+  =/  doc/dock  [king.tie.sta dap.bol]
+  [ost.bol %poke /ask-king doc %tiebout-action [%notify not]]~
```

Comment:
Ah, so kings route your notifications for you? Is this because we don't want any old ship to access our push endpoint? idk about loading it all onto ~dopzod, we may want a dedicated ship for this, but the load probably isn't *that* big?